lib/src/wiggle_abi/dictionary_impl.rs

```diff
@@ -9,20 +12,63 @@ use {
     wiggle::GuestPtr,
 };
 
+use memoize::memoize;
+
+#[memoize]
+fn read_json_file(file: String) -> serde_json::Map<String, serde_json::Value> {
+    let data = fs::read_to_string(file).expect("Unable to read file");
+    let json: serde_json::Value = serde_json::from_str(&data).expect("JSON was not well-formatted");
+    let obj = json.as_object().expect("Expected the JSON to be an Object");
+    obj.clone()
+}
+
 impl FastlyDictionary for Session {
-    #[allow(unused_variables)] // FIXME: Remove this directive once implemented.
     fn open(&mut self, name: &GuestPtr<str>) -> Result<DictionaryHandle, Error> {
-        Err(Error::NotAvailable("Dictionary lookup"))
+        let dicts = self.dictionaries.as_ref();
+        let name = name.as_str()?.to_owned();
+        match dicts.get(&name) {
+            Some(dictionary) => Ok(dictionary.id),
+            None => Err(Error::UnknownDictionary(name)),
+        }
     }
 
-    #[allow(unused_variables)] // FIXME: Remove this directive once implemented.
     fn get(
         &mut self,
         dictionary: DictionaryHandle,
         key: &GuestPtr<str>,
         buf: &GuestPtr<u8>,
         buf_len: u32,
     ) -> Result<u32, Error> {
-        Err(Error::NotAvailable("Dictionary lookup"))
+        let dicts = self.dictionaries_by_id.as_ref();
+        let key = key.as_str()?.to_owned();
+        match dicts.get(&dictionary) {
+            Some(dict) => {
+                let file = dict.file.clone();
+                let obj = read_json_file(file);
+                if !obj.contains_key(&key) {
+                    return Err(Error::UnknownDictionaryItem(key));
+                }
+                let item = obj.get(&key).unwrap(); // Safe to do due to `!obj.contains_key(&key)` above
+                let item = serde_json::to_string(item).unwrap();
+                let item_bytes = item.as_bytes();
+
+                if item_bytes.len() > buf_len as usize {
+                    // Write out the number of bytes necessary to fit this dictionary_item, or zero on overflow to
+                    // signal an error condition.
+                    buf.write(item_bytes.len().try_into().unwrap_or(0))?;
+                    return Err(Error::BufferLengthError {
+                        buf: "dictionary_item",
+                        len: "dictionary_item_max_len",
```

Comment:
It's not possible via the Fastly API to create a dictionary item that is larger than `dictionary_item_max_len`, which means this is only a possible runtime error when using Viceroy.

Should we have Viceroy check all the dictionary items are below `dictionary_item_max_len` prior to starting up the server? It would change this from being a runtime error to a compile time error, giving users a shorter feedback loop

Reply:
I've added this in [`a61a697` (#61)](https://github.com/fastly/Viceroy/pull/61/commits/a61a6978eae28d216dc014a30ce982c6cc62e7eb) 👍